Client/Frontend/Browser/Search/BraveSearchManager.swift

```diff
@@ -0,0 +1,280 @@
+        
+        let headers = HTTPCookie.requestHeaderFields(with: domainCookies)
+        headers.forEach {
+            request.setValue($0.value, forHTTPHeaderField: $0.key)
+        }
+        
+        let session = URLSession(configuration: .ephemeral, delegate: self, delegateQueue: .main)
+        
+        // Important, URLSessionDelegate must have been implemented here
+        // to handle request authentication.
+        session
+            .dataTaskPublisher(for: request)
+            .tryMap { output -> Data in
+                guard let response = output.response as? HTTPURLResponse,
+                      response.statusCode >= 200 && response.statusCode < 300 else {
+                    throw "Invalid response"
+                }
+     
+                return output.data
+            }
+            .decode(type: BackupQuery.self, decoder: JSONDecoder())
+            .sink(receiveCompletion: { status in
+                switch status {
+                case .failure(let error):
+                    log.error("shouldUseFallback error: \(error)")
+                    // No subsequent call to backup search engine is made in case of error.
+                    // The pending status has to be cancelled.
+                    completion(nil)
+                case .finished:
+                    // Completion is called on `receiveValue`.
+                    break
+                }
+            }, receiveValue: { canAnswer in
+                completion(canAnswer)
+            })
+            .store(in: &cancellables)
+        
+        session.finishTasksAndInvalidate()
+    }
+    
+    /// Perform a backup search using an alternative search engine, gets results back as html source.
+    func backupSearch(with backupQuery: BackupQuery,
+                      completion: @escaping (String) -> Void) {
+        
+        guard var components = URLComponents(string: fallbackProviderURLString) else { return }
+        
+        var queryItems: [URLQueryItem] = [
+            .init(name: "q", value: query.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed))]
+        
+        if let language = backupQuery.language {
+            queryItems.append(.init(name: "hl", value: language))
+        }
+        
+        if let country = backupQuery.country {
+            queryItems.append(.init(name: "gl", value: country))
+        }
+        
+        components.queryItems = queryItems
+        
+        guard let url = components.url else { return }
+        var request = URLRequest(url: url, timeoutInterval: 5)
+        
+        // Must be set, without it the fallback results may be not retrieved correctly.
+        request.addValue(UserAgent.userAgentForDesktopMode, forHTTPHeaderField: "User-Agent")
+        
+        request.addValue("text/html;charset=UTF-8, text/plain;charset=UTF-8",
+                         forHTTPHeaderField: "Accept")
+        
+        URLSession(configuration: .ephemeral)
```

Comment:
my understanding is that this prevents cache/cookies from being written to disk (https://developer.apple.com/documentation/foundation/urlsessionconfiguration/1410529-ephemeral), but google can still set/read cookies that are persisted in memory. i think we want to prevent cookies/cache from being used at all here.

Reply:
i might be misunderstanding but i don't see where `finishTasksAndInvalifate` is called for the Google request's session, only the session for the Brave search API.

Reply:
correct, i have to add it for the Google request as well